mlir/lib/Driver/Timer.hpp

```diff
@@ -0,0 +1,145 @@
+    }
+
+    void stop() noexcept
+    {
+        if (debug_timer && running) {
+            stop_time_ = std::chrono::steady_clock::now();
+            running = false;
+        }
+    }
+
+    auto elapsed() noexcept
+    {
+        if (debug_timer) {
+            if (running) {
+                stop();
+            }
+            return std::chrono::duration_cast<std::chrono::nanoseconds>(stop_time_ - start_time_);
+        }
+        else {
+            return std::chrono::nanoseconds(0);
+        }
+    }
+
+    void print(const std::string &name) noexcept
+    {
+        // Convert nanoseconds (long) to milliseconds (double)
+        const auto ms = static_cast<double>(elapsed().count()) / 1e6;
+        // Get the hash of id as there is no conversion from id to size_t (or string)
+        const auto id = std::hash<std::thread::id>{}(std::this_thread::get_id());
+
+        std::cerr << "[TIMER] Running " << name << " in " << ms << "ms";
+        std::cerr << " (thread_id = " << std::to_string(id) << ")" << std::endl;
+    }
+
+    void store(const std::string &name, const std::string &key,
+               const std::filesystem::path &file_path)
+    {
+        // Convert nanoseconds (long) to milliseconds (double)
+        const auto ms = static_cast<double>(elapsed().count()) / 1e6;
+        // Get the hash of id as there is no conversion from id to size_t (or string)
+        const auto id = std::hash<std::thread::id>{}(std::this_thread::get_id());
+        // Create YAML headers with key and thread-id conditionally
+        const auto header = key + " (thread_id=" + std::to_string(id) + ")";
+
+        if (!std::filesystem::exists(file_path)) {
+            std::ofstream ofile(file_path);
+            assert(ofile.is_open() && "Invalid file to store timer results");
+            ofile << header << ":" << std::endl;
+            ofile << "  - " << name << ": " << ms << "ms" << std::endl;
+            ofile.close();
+            return;
+        }
+        // else
+        // First, check if the header is in the file
+        std::ifstream ifile(file_path);
+        assert(ifile.is_open() && "Invalid file to store timer results");
+        std::string line;
+        bool add_header = true;
+        while (add_header && std::getline(ifile, line)) {
+            if (line.find(header) != std::string::npos) {
+                add_header = false;
+            }
+        }
+        ifile.close();
+
+        // Second, update the file
+        std::ofstream ofile(file_path, std::ios::app);
+        assert(ofile.is_open() && "Invalid file to store timer results");
+        if (add_header) {
+            ofile << header << ":" << std::endl;
+        }
+        ofile << "  - " << name << ": " << ms << "ms" << std::endl;
+        ofile.close();
+    }
+
+    void dump(const std::string &name, const std::string &key)
+    {
+        if (!debug_timer) {
+            return;
+        }
+
+        char *file = getenv("DEBUG_TIMER_RESULT_PATH");
```

Comment:
I'm a fan of avoiding verbosity if possible when manual input is required, so even shortening by one word might be nice:
```suggestion
        char *file = getenv("DEBUG_TIMER_PATH");
```

Reply:
Update it to `DEBUG_RESULTS_FILE` as it contains both timing and program-size results.